langkit/compile_context.py

```diff
@@ -882,6 +882,27 @@ def compute_types(self):
                  value_names=[self.grammar_rule_api_name(n)
                               for n in self.grammar.user_defined_rules])
 
+    def compute_optional_field_info(self):
+        """
+        For every parse field, find out if it is an optional field or not, i.e.
+        whether it is ever produced from an Opt parser in the user grammar.
+        """
+        from langkit.parsers import Opt
+
+        all_parse_fields = [
+            field
+            for node_type in self.astnode_types
+            for field in node_type.get_parse_fields(include_inherited=False)
+        ]
+
+        for field in all_parse_fields:
+            field._is_optional = False
+            for parser in field.parsers_from_transform:
+                # If parser is an Opt parser and is not set to produce an enum
+                # alternative, it means that field is optional.
+                if isinstance(parser, Opt) and not parser._booleanize:
+                    field._is_optional = True
```

Comment:
If I understand correctly, the goal is not really to determine if an `Opt` parser is involved, but rather if the parser can return a null node, right? If so, I think this analysis should be much more involved. For instance when you have an `Or` subparser, you then need to check if any subparser for this `Or` can itself return a null node. You must also check for `Null` parsers, recurse in `Defer` parsers, etc.